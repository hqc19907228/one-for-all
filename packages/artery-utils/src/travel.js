import { Map } from 'immutable';
import { isImmutable } from 'immutable';
import { fromJS } from 'immutable';
import walk from './raw/walk';

const NODE_TYPE_TRAVELER_MAP = {
  'html-element': 'htmlNode',
  'react-component': 'reactComponentNode',
  'loop-container': 'loopContainerNode',
  'composed-node': 'composedNode',
  'ref-node': 'refNode',
  'jsx-node': 'jsxNode',
  'route-node': 'routeNode',
};

export default function travel(schemaNode, Visitors) {
  const node = fromJS(schemaNode);
  const _newRoot = walk(node, (newRoot, keyPath) => {
    if (!newRoot) {
      newRoot = node;
    }

    const nodeType = node.getIn(keyPath.concat(['type']));
    const currentNode = node.getIn(keyPath).toJS();
    const newNode = Visitors[NODE_TYPE_TRAVELER_MAP[nodeType]]?.(currentNode);
    if (newNode !== undefined) {
      newRoot = newRoot.setIn(keyPath, newNode);
      if (!isImmutable(newRoot)) {
        newRoot = Map(newRoot);
      }
    }

    return newRoot;
  });

  return _newRoot.toJS();
}
