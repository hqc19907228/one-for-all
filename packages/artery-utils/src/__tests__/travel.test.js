import schemaNode from '../raw/__tests__/fixtures/node';
import travel from '../travel';

test('travel_call_Visitors', () => {
  const newRoot = travel(schemaNode, {
    htmlNode: (currentNode) => {
      currentNode.id = `new__${currentNode.id}`;
      return currentNode;
    },
    reactComponentNode: (currentNode) => {
      currentNode.id = `new__${currentNode.id}`;
      return currentNode;
    },
    loopContainerNode: (currentNode) => {
      currentNode.id = `new__${currentNode.id}`;
      return currentNode;
    },
    composedNode: (currentNode) => {
      currentNode.id = `new__${currentNode.id}`;
      return currentNode;
    },
    refNode: (currentNode) => {
      currentNode.id = `new__${currentNode.id}`;
      return currentNode;
    },
    jsxNode: (currentNode) => {
      currentNode.id = `new__${currentNode.id}`;
      return currentNode;
    },
    routeNode: (currentNode) => {
      currentNode.id = `new__${currentNode.id}`;
      return currentNode;
    },
  });

  expect(newRoot).toMatchSnapshot();
});
