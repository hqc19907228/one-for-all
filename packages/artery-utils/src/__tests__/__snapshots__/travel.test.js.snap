// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`travel_call_Visitors 1`] = `
Object {
  "children": Array [
    Object {
      "id": "new__ref-schema",
      "schemaID": "whatever",
      "type": "ref-node",
    },
    Object {
      "children": Array [
        Object {
          "exportName": "TodoInput",
          "id": "new__todo-input",
          "packageName": "todo-app",
          "packageVersion": "whatever",
          "props": Object {
            "onEnter": Object {
              "func": Object {
                "args": "value",
                "body": "
                const title = value.trim();
                if (!title) {
                  return;
                }

                this.apiStates.postTodo.fetch({ body: { title }}, () => {
                  this.apiStates[\\"全部待办列表\\"].refresh();
                })
                ",
                "type": "raw",
              },
              "type": "functional_property",
            },
          },
          "supportStateExposure": true,
          "type": "react-component",
        },
      ],
      "id": "new__todo-input-container",
      "name": "div",
      "type": "html-element",
    },
    Object {
      "id": "new__todo-list-loop-composedNode",
      "iterableState": Object {
        "convertor": Object {
          "expression": "state",
          "type": "state_convert_expression",
        },
        "fallback": Array [],
        "stateID": "全部待办列表",
        "type": "api_result_property",
      },
      "loopKey": "id",
      "node": Object {
        "children": Array [],
        "id": "new__compose-node-container",
        "nodes": Array [
          Object {
            "id": "new__todo-toggle",
            "name": "input",
            "props": Object {
              "checked": Object {
                "type": "constant_property",
                "value": true,
              },
              "onChange": Object {
                "func": Object {
                  "args": "e",
                  "body": "
                  this.apiStates['更新待办'].fetch(
                    { params: { todoId: e.target.dataset.id } },
                    () => this.apiStates[\\"全部待办列表\\"].refresh()
                    )",
                  "type": "raw",
                },
                "type": "functional_property",
              },
              "type": Object {
                "type": "constant_property",
                "value": "checkbox",
              },
            },
            "toProps": Object {
              "args": "state",
              "body": "
              return {
                'data-id': state.id,
                checked: state.status === \\"working\\" ? false : true,
              }",
              "type": "to_props_function_spec",
            },
            "type": "html-element",
          },
          Object {
            "id": "new__todo-title",
            "name": "span",
            "toProps": Object {
              "args": "state",
              "body": "
                return {
                  children: state.title,
                }
                ",
              "type": "to_props_function_spec",
            },
            "type": "html-element",
          },
          Object {
            "id": "new__todo-delete",
            "name": "button",
            "props": Object {
              "children": Object {
                "type": "constant_property",
                "value": "X",
              },
              "onClick": Object {
                "func": Object {
                  "args": "e",
                  "body": "
                    this.apiStates['删除待办'].fetch({ params: { todoId: e.target.id }}, () => {
                      this.apiStates[\\"全部待办列表\\"].refresh()
                    })
                    ",
                  "type": "raw",
                },
                "type": "functional_property",
              },
              "style": Object {
                "type": "constant_property",
                "value": Object {
                  "display": "inline-block",
                  "float": "right",
                  "width": "50px",
                },
              },
            },
            "toProps": Object {
              "args": "state",
              "body": "
                return {
                  id: state.id,
                }
                ",
              "type": "to_props_function_spec",
            },
            "type": "html-element",
          },
        ],
        "outLayer": Object {
          "id": "todo-item-outLayer",
          "name": "div",
          "props": Object {
            "style": Object {
              "type": "constant_property",
              "value": Object {
                "marginBottom": "8px",
              },
            },
          },
          "type": "html-element",
        },
        "type": "composed-node",
      },
      "props": Object {},
      "type": "loop-container",
    },
    Object {
      "children": Array [
        Object {
          "exportName": "TodoFilter",
          "id": "new__todo-filter",
          "packageName": "todo-app",
          "packageVersion": "whatever",
          "props": Object {
            "onFetchStatus": Object {
              "stateID": "todoStatus",
              "type": "api_invoke_property",
            },
            "onToggleStatus": Object {
              "paramsBuilder": Object {
                "args": "status",
                "body": "
                  return { params: { status } };
                  ",
                "type": "param_builder_func_spec",
              },
              "stateID": "全部待办列表",
              "type": "api_invoke_property",
            },
          },
          "type": "react-component",
        },
      ],
      "id": "new__footer",
      "name": "div",
      "props": Object {
        "style": Object {
          "type": "constant_property",
          "value": Object {
            "borderTop": "1px solid #eee",
            "display": "flex",
            "justifyContent": "space-between",
            "paddingTop": "16px",
          },
        },
      },
      "type": "html-element",
    },
  ],
  "id": "new__container",
  "lifecycleHooks": Object {
    "didMount": Object {
      "args": "",
      "body": "this.apiStates['全部待办列表'].fetch();",
      "type": "lifecycle_hook_func_spec",
    },
  },
  "name": "div",
  "props": Object {
    "id": Object {
      "type": "constant_property",
      "value": "container",
    },
    "style": Object {
      "type": "constant_property",
      "value": Object {
        "borderRadius": "2px",
        "boxShadow": "blue 0px 0px 0px 2px inset, rgb(255, 255, 255) 10px -10px 0px -3px, rgb(31, 193, 27) 10px -10px, rgb(255, 255, 255) 20px -20px 0px -3px, rgb(255, 217, 19) 20px -20px, rgb(255, 255, 255) 30px -30px 0px -3px, rgb(255, 156, 85) 30px -30px, rgb(255, 255, 255) 40px -40px 0px -3px, rgb(255, 85, 85) 40px -40px",
        "margin": "auto",
        "marginTop": "100px",
        "padding": "20px",
        "width": "500px",
      },
    },
  },
  "type": "html-element",
}
`;
